walkeeper/src/safekeeper.rs

```diff
@@ -0,0 +1,444 @@
+    vote_given: u64, // fixme u64 due to padding
+    /// Safekeeper's log position, to let proposer choose the most advanced one
+    epoch: Term,
+    flush_lsn: Lsn,
+    restart_lsn: Lsn,
+}
+
+/// Request with WAL message sent from proposer to safekeeper. Along the way it
+/// announces 1) successfull election (with VCL); 2) commit_lsn.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AppendRequest {
+    h: AppendRequestHeader,
+    wal_data: Bytes,
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AppendRequestHeader {
+    term: Term,
+    /// volume commit LSN
+    vcl: Lsn,
+    /// start position of message in WAL
+    begin_lsn: Lsn,
+    /// end position of message in WAL
+    end_lsn: Lsn,
+    /// LSN committed by quorum of safekeepers
+    commit_lsn: Lsn,
+    /// restart LSN position  (minimal LSN which may be needed by proposer to perform recovery)
+    restart_lsn: Lsn,
+    // only for logging/debugging
+    proposer_uuid: PgUuid,
+}
+
+/// Report safekeeper state to proposer
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct AppendResponse {
+    // Current term of the safekeeper; if it is higher than proposer's, the
+    // compute is out of date.
+    pub term: Term,
+    pub epoch: Term,
+    // NOTE: this is physical end of wal on safekeeper; currently it doesn't
+    // make much sense without taking epoch into account, as history can be
+    // diverged.
+    pub flush_lsn: Lsn,
+    pub hs_feedback: HotStandbyFeedback,
+}
+
+/// Proposer -> Acceptor messages
+#[derive(Debug)]
+pub enum ProposerAcceptorMessage {
+    Greeting(ProposerGreeting),
+    VoteRequest(VoteRequest),
+    AppendRequest(AppendRequest),
+}
+
+fn des_stream_le<T: LeSer, S: io::Read>(stream: &mut S) -> Result<T> {
+    // As the trait bound implies, this always encodes little-endian.
+    Ok(T::des_from(stream)?)
+}
+
+impl ProposerAcceptorMessage {
+    /// Parse proposer message.
+    pub fn parse(msg: Bytes) -> Result<ProposerAcceptorMessage> {
+        // xxx this is not efficient but easy to work with bin_ser
+        let mut stream = msg.reader();
+        // u64 is here to avoid padding; it will be removed once we stop packing C structs into the wire as is
+        let tag = stream.read_u64::<LittleEndian>()? as u8 as char;
+        match tag {
+            'g' => {
+                let msg = des_stream_le::<ProposerGreeting, Reader<Bytes>>(&mut stream)?;
+                Ok(ProposerAcceptorMessage::Greeting(msg))
+            }
+            'v' => {
+                let msg = des_stream_le::<VoteRequest, Reader<Bytes>>(&mut stream)?;
+                Ok(ProposerAcceptorMessage::VoteRequest(msg))
+            }
+            'a' => {
+                // read header followed by wal data
+                let hdr = des_stream_le::<AppendRequestHeader, Reader<Bytes>>(&mut stream)?;
+                let rec_size = hdr.end_lsn.checked_sub(hdr.begin_lsn).unwrap().0 as usize;
```

Comment:
This panics if the message was garbage. That doesn't seem nice, the safekeepers are multi-tenant, right? I'm not sure how careful we are about that in general, maybe it's a todo to tighten up error handling to not panic the whole process if something's wrong with one tenant..

Reply:
AFAIK by default without setting panic to abort in Cargo.toml it will just bring down current thread, so since we use thread per connection model  it is not that bad, but having more granular errors is a good thing anyway

Reply:
Yeah, let's be more careful here.